### PR TITLE
Temporarily pin versions for integration tests to `8.16.0`

### PR DESCRIPTION
**Pull request by ycombinator**
While the Agent version has been set to `8.17.0` on the `8.x` branch, it's not yet possible to create `8.17.0-SNAPSHOT` deployments in ESS.  As such, Agent integration tests — which use deployments in ESS — will fail on the `8.x` branch.

To work around this problem, this PR temporarily builds Elastic Agent with the `8.16.0` version and uses this build in integration tests.  Likewise, it asks the integration tests to spin up `8.16.0-SNAPSHOT` deployments in ESS.

---

We should revert this PR once it becomes possible to create `8.17.0-SNAPSHOT` deployments in ESS.

### CODE REVIEW

**Review by jlind23 — APPROVED** [review]

**Comment by jlind23** [comment]
buildkite test it

**Comment by pierrehilbert** [comment]
Having an automatic fallback could be an idea to make our life easier but it could also make us miss that we are not running on the expected version and therefore miss something.

**Comment by ycombinator** [comment]
Serverless integration tests are still failing on this PR with the error message:

```
Error: error running test: failed to create ESS cloud for version 8.17.0-SNAPSHOT: failed to create: (clusters.cluster_invalid_plan) Version [8.17.0-SNAPSHOT] is not available. Please use one of the available versions
```

@pchila Can you spot which env var I missed setting?

File: .buildkite/scripts/steps/k8s-extended-tests.sh [hunk 1]
```diff
@@ -25,8 +25,8 @@ else
   exit 10
 fi
 
-DEV=true SNAPSHOT=true EXTERNAL=true PACKAGES=docker mage -v package
-TEST_INTEG_CLEAN_ON_EXIT=true INSTANCE_PROVISIONER=kind STACK_PROVISIONER=stateful SNAPSHOT=true mage integration:kubernetesMatrix
+AGENT_PACKAGE_VERSION="8.16.0" DEV=true SNAPSHOT=true EXTERNAL=true PACKAGES=docker mage -v package
+AGENT_PACKAGE_VERSION="8.16.0-SNAPSHOT" TEST_INTEG_CLEAN_ON_EXIT=true INSTANCE_PROVISIONER=kind STACK_PROVISIONER=stateful SNAPSHOT=true mage integration:kubernetesMatrix
```
Comment (pchila) [inline .buildkite/scripts/steps/k8s-extended-tests.sh:29]:
```suggestion
AGENT_PACKAGE_VERSION="8.16.0-SNAPSHOT" AGENT_STACK_VERSION="8.16.0-SNAPSHOT" TEST_INTEG_CLEAN_ON_EXIT=true INSTANCE_PROVISIONER=kind STACK_PROVISIONER=stateful SNAPSHOT=true mage integration:kubernetesMatrix
```

**Comment by pchila** [comment]
> Integration tests and extended runtime leak tests are still failing on this PR with the error message:
> 
> ```
> Error: error running test: failed to create ESS cloud for version 8.17.0-SNAPSHOT: failed to create: (clusters.cluster_invalid_plan) Version [8.17.0-SNAPSHOT] is not available. Please use one of the available versions
> ```
> 
> @pchila Can you spot which env var I missed setting?

the problem is in the version set in package.version

https://github.com/elastic/elastic-agent/blob/8.x/.package-version

we need an extra override

File: .buildkite/scripts/steps/integration_tests.sh [hunk 2]
```diff
@@ -19,7 +19,7 @@ fi
 
 # Run integration tests
 set +e
-AGENT_STACK_VERSION="${STACK_VERSION}" TEST_INTEG_CLEAN_ON_EXIT=true  STACK_PROVISIONER="$STACK_PROVISIONER" SNAPSHOT=true mage $MAGE_TARGET $MAGE_SUBTARGET
+AGENT_VERSION="8.16.0-SNAPSHOT" AGENT_STACK_VERSION="${STACK_VERSION}" TEST_INTEG_CLEAN_ON_EXIT=true  STACK_PROVISIONER="$STACK_PROVISIONER" SNAPSHOT=true mage $MAGE_TARGET $MAGE_SUBTARGET
```
Comment (pchila) [inline .buildkite/scripts/steps/integration_tests.sh:22]:
```suggestion
AGENT_VERSION="8.16.0-SNAPSHOT" AGENT_STACK_VERSION="8.16.0" TEST_INTEG_CLEAN_ON_EXIT=true  STACK_PROVISIONER="$STACK_PROVISIONER" SNAPSHOT=true mage $MAGE_TARGET $MAGE_SUBTARGET
```

Comment (pchila) [reply .buildkite/scripts/steps/integration_tests.sh]: This is slightly brutal but it should nullify the 8.17.0 version we read from package.version

File: .buildkite/scripts/steps/k8s-extended-tests.sh [hunk 3]
```diff
@@ -25,8 +25,8 @@ else
   exit 10
 fi
 
-DEV=true SNAPSHOT=true EXTERNAL=true PACKAGES=docker mage -v package
-TEST_INTEG_CLEAN_ON_EXIT=true INSTANCE_PROVISIONER=kind STACK_PROVISIONER=stateful SNAPSHOT=true mage integration:kubernetesMatrix
+AGENT_PACKAGE_VERSION="8.16.0" DEV=true SNAPSHOT=true EXTERNAL=true PACKAGES=docker mage -v package
+AGENT_PACKAGE_VERSION="8.16.0-SNAPSHOT" AGENT_STACK_VERSION="8.16.0-SNAPSHOT" TEST_INTEG_CLEAN_ON_EXIT=true INSTANCE_PROVISIONER=kind STACK_PROVISIONER=stateful SNAPSHOT=true mage integration:kubernetesMatrix
```
Comment (pchila) [inline .buildkite/scripts/steps/k8s-extended-tests.sh:29]:
```suggestion
AGENT_VERSION="8.16.0-SNAPSHOT" AGENT_STACK_VERSION="8.16.0-SNAPSHOT" TEST_INTEG_CLEAN_ON_EXIT=true INSTANCE_PROVISIONER=kind STACK_PROVISIONER=stateful SNAPSHOT=true mage integration:kubernetesMatrix
```

File: .buildkite/scripts/steps/integration_tests.sh [hunk 4]
```diff
@@ -19,7 +19,7 @@ fi
 
 # Run integration tests
 set +e
-AGENT_STACK_VERSION="${STACK_VERSION}" TEST_INTEG_CLEAN_ON_EXIT=true  STACK_PROVISIONER="$STACK_PROVISIONER" SNAPSHOT=true mage $MAGE_TARGET $MAGE_SUBTARGET
+AGENT_VERSION="8.16.0-SNAPSHOT" AGENT_STACK_VERSION="8.16.0" TEST_INTEG_CLEAN_ON_EXIT=true  STACK_PROVISIONER="$STACK_PROVISIONER" SNAPSHOT=true mage $MAGE_TARGET $MAGE_SUBTARGET
```
Comment (pchila) [inline .buildkite/scripts/steps/integration_tests.sh:22]:
```suggestion
AGENT_VERSION="8.16.0-SNAPSHOT" AGENT_STACK_VERSION="8.16.0-SNAPSHOT" TEST_INTEG_CLEAN_ON_EXIT=true  STACK_PROVISIONER="$STACK_PROVISIONER" SNAPSHOT=true mage $MAGE_TARGET $MAGE_SUBTARGET
```

**Review by pierrehilbert — APPROVED** [review]

**Comment by cmacknz** [comment]
> this PR temporarily builds Elastic Agent with the 8.16.0 version and uses this build in integration tests. Likewise, it asks the integration tests to spin up 8.16.0-SNAPSHOT deployments in ESS.

Only the second part (creating 8.16.0-SNAPSHOT deployments in ESS) should be necessary now. Fleet Server should let agents that are one minor newer connect when fleet server is a snapshot. It didn't handle the major version bump for 9.0 though, which is where the need to change the package version came back from. We should be fine on that front for a while.

**Pull request merged**